File: src/main/java/essentials/internal/thread/AutoRollback.java
package essentials.internal.thread;

import arc.files.Fi;
import arc.struct.Seq;
import essentials.internal.CrashReport;
import essentials.internal.Log;
import mindustry.core.GameState;
import mindustry.gen.Call;
import mindustry.gen.Groups;
import mindustry.gen.Playerc;
import mindustry.io.SaveIO;

import java.util.TimerTask;

import static essentials.Main.config;
import static mindustry.Vars.*;

public class AutoRollback extends TimerTask {
    public void save() {
        try {
            Fi file = saveDirectory.child(config.slotNumber() + "." + saveExtension);
            if (state.is(GameState.State.playing)) SaveIO.save(file);
        } catch (Exception e) {
            new CrashReport(e);
        }
    }

    public void load() {
        Seq<Playerc> players = new Seq<>();
        for (Playerc p : Groups.player) {
            players.add(p);
            p.dead();
        }

        logic.reset();

        Call.onWorldDataBegin();

        try {
            Fi file = saveDirectory.child(config.slotNumber() + "." + saveExtension);
            SaveIO.load(file);

            logic.play();

            for (Playerc p : players) {
                if (p.con() == null) continue;

                p.reset();
                if (state.rules.pvp) {
                    p.team(netServer.assignTeam(p, new Seq.SeqIterable<>(players)));
                }
                netServer.sendWorldData(p);
            }
        } catch (SaveIO.SaveException e) {
            new CrashReport(e);
        }
        Log.info("Map rollbacked.");
        /*new Thread(() -> {
            try {
                float orignal = state.rules.respawnTime;
                state.rules.respawnTime = 0f;
                Call.onSetRules(state.rules);
                sleep(3000);
                state.rules.respawnTime = orignal;
                Call.onSetRules(state.rules);
            } catch (InterruptedException ignored) {
                Thread.currentThread().interrupt();
            }

        }).start();*/
        if (state.is(GameState.State.playing)) Call.sendMessage("[green]Map rollbacked.");
    }

    @Override
    public void run() {
        Thread.currentThread().setName("Essential Auto rollback thread");
        save();
    }
}

File: src/main/kotlin/essentials/thread/AutoRollback.kt
package essentials.thread

import arc.struct.Seq
import arc.struct.Seq.SeqIterable
import essentials.Main.Companion.configs
import essentials.internal.CrashReport
import essentials.internal.Log
import mindustry.Vars
import mindustry.Vars.netServer
import mindustry.core.GameState
import mindustry.game.Gamemode
import mindustry.gen.Call
import mindustry.gen.Groups
import mindustry.gen.Player
import mindustry.io.SaveIO
import mindustry.io.SaveIO.SaveException
import mindustry.maps.Map
import java.util.*


class AutoRollback : TimerTask() {
    fun save() {
        try {
            val file = Vars.saveDirectory.child(configs.slotNumber.toString() + "." + Vars.saveExtension)
            if (Vars.state.`is`(GameState.State.playing)) SaveIO.save(file)
        } catch (e: Exception) {
            CrashReport(e)
        }
    }

    fun load() {
        val players = Seq<Player>()
        for (p in Groups.player) {
            players.add(p)
            p.dead()
        }
        Vars.logic.reset()
        Call.worldDataBegin()
        try {
            val file = Vars.saveDirectory.child(configs.slotNumber.toString() + "." + Vars.saveExtension)
            SaveIO.load(file)
            Vars.logic.play()
            for (p in players) {
                if (p.con() == null) continue
                p.reset()
                if (Vars.state.rules.pvp) {
                    p.team(netServer.assignTeam(p, SeqIterable(players)))
                }
                Vars.netServer.sendWorldData(p)
            }
        } catch (e: SaveException) {
            CrashReport(e)
        }
        Log.info("Map rollbacked.")
        if (Vars.state.`is`(GameState.State.playing)) Call.sendMessage("[green]Map rollbacked.")
    }

    fun load(map: Map?) {
        val players = Seq<Player>()
        for (p in Groups.player) {
            players.add(p)
            p.dead()
        }
        Vars.logic.reset()
        Call.worldDataBegin()
        try {
            Vars.world.loadMap(map, map!!.applyRules(Gamemode.survival))
            //SaveIO.load(map.file);
            Vars.logic.play()
            for (p in players) {
                if (p.con() == null) continue
                p.reset()
                if (Vars.state.rules.pvp) {
                    p.team(netServer.assignTeam(p, SeqIterable(players)))
                }
                Vars.netServer.sendWorldData(p)
            }
        } catch (e: SaveException) {
            CrashReport(e)
        }
    }

    override fun run() {
        Thread.currentThread().name = "Essential Auto rollback thread"
        save()
    }
}